### ISSUE TITLE
`isinstance(foo, NoneType)` does not narrow `foo` as `None`

### ISSUE THREAD

**Issue by Jasha10** [microsoft/pyright#4402]
**Describe the bug**
With a variable `foo` typed as `T | None`, the statement `isinstance(foo, NoneType)` does not narrow `foo` to be `None`.

**To Reproduce**
```python
# tmp.py
from random import getrandbits
from types import NoneType
from typing_extensions import assert_never

random_bool = bool(getrandbits(1))

foo: int | None = 123 if random_bool else None

if isinstance(foo, int):
    ...
elif isinstance(foo, NoneType):
    ...
else:
    assert_never(foo)
```
```text
$ pyright tmp.py
...
  .../tmp.py:15:18 - error: Argument of type "None" cannot be assigned to parameter "__arg" of type "Never" in function "assert_never"
    Type "None" cannot be assigned to type "Never" (reportGeneralTypeIssues)
```

**Expected behavior**
I would not expect pyright to give an error; I'd expect pyright to narrow `foo` as `None` in the second arm of the if-elif block above.

**Additional context**
I understand that the annotation `foo: int | NoneType` might be more technically correct than `foo: int | None`. However, using the former annotation violates convention and causes pyright to spit out a different error (namely `tmp.py:8:23 - error: Expression of type "Literal[123] | None" cannot be assigned to declared type "int | NoneType"`).

I discovered this bug when trying to use a match statement as so:
```python
match foo:
    case int():
        ...
    case NoneType():
        ...
    case _:
        assert_never(foo)  # pyright err: Argument of type "None" cannot be assigned...
```

**Comment by erictraut** [comment]
`NoneType` was not exposed until 3.10. Prior to that, the standard way to check for `None` using `isinstance` was `isinstance(foo, type(None))`. Both pyright and mypy support this.

I think it's reasonable to expect `NoneType` to work here as well.

**Comment by erictraut** [comment]
This will be included in the next release.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.288, which I just published. It will also be included in a future release of pylance.
